src/blocks/keyboard_layout.rs

```diff
@@ -318,3 +330,76 @@ trait LocaleBusInterface {
     #[dbus_proxy(property, name = "X11Variant")]
     fn variant(&self) -> zbus::Result<String>;
 }
+
+#[dbus_proxy(
+    interface = "ru.gentoo.kbdd",
+    default_service = "ru.gentoo.KbddService",
+    default_path = "/ru/gentoo/KbddService"
+)]
+trait KbddBusInterface {
+    #[dbus_proxy(signal, name = "layoutNameChanged")]
+    fn layout_updated(&self, layout: String) -> zbus::Result<()>;
+
+    #[dbus_proxy(name = "getCurrentLayout")]
+    fn current_layout_index(&self) -> zbus::Result<u32>;
+
+    #[dbus_proxy(name = "getLayoutName")]
+    fn current_layout(&self, layout_id: u32) -> zbus::Result<String>;
+}
+
+struct KbddBus {
+    stream: layoutNameChangedStream<'static>,
+    current_layout: String,
+    re_layout_name: Regex,
+}
+
+impl KbddBus {
+    async fn new() -> Result<Self> {
+        let conn = new_dbus_connection().await?;
+        let proxy = KbddBusInterfaceProxy::builder(&conn)
+            .cache_properties(zbus::CacheProperties::No)
+            .build()
+            .await
+            .error("Failed to create KbddBusInterfaceProxy")?;
+        let stream = proxy
+            .receive_layout_updated()
+            .await
+            .error("Failed to monitor kbdd interface")?;
+        let layout_index = proxy
+            .current_layout_index()
+            .await
+            .error("Failed to get current layout index from kbdd")?;
+        let current_layout = proxy
+            .current_layout(layout_index)
+            .await
+            .error("Failed to get current layout from kbdd")?;
+        Ok(Self {
+            stream,
+            current_layout,
+            re_layout_name: Regex::new(r#"(.*) \((.*)\)"#).unwrap(), //to split layout from variant in string like 'English (US)'
+        })
+    }
+}
+
+#[async_trait]
+impl Backend for KbddBus {
+    async fn get_info(&mut self) -> Result<Info> {
+        let (layout, variant) = match self.re_layout_name.captures(&self.current_layout) {
+            Some(caps) => (
+                caps.get(1).unwrap().as_str().to_string(),
+                Some(caps.get(2).unwrap().as_str().to_string()),
+            ),
+            //fallback, although this should not happen
+            None => (self.current_layout.to_string(), None),
+        };
+        Ok(Info { layout, variant })
+    }
+
+    async fn wait_for_chagne(&mut self) -> Result<()> {
+        if let Some(event) = self.stream.next().await {
```

Comment:
Perhaps it's better to throw an error if `.next()` returns `None`?

Reply:
I agree, probably it's best if this is treated as error. I've updated the PR.